client/client.go

```diff
@@ -136,77 +238,107 @@ func decodeHeader(header []byte) (code []byte, pt uint32, length int) {
 	return
 }
 
-func (client *Client) reconnect(err error) error {
-	if client.conn != nil {
-		return nil
+func (client *Client) grabReconnectState() (success bool) {
+	client.reconnectLock.Lock()
+	defer client.reconnectLock.Unlock()
+	if client.reconnectingActive { // another thread is already reconnecting to server, this thread will exit
+		return false
 	}
+	client.reconnectingActive = true // I am first to attempt reconnection
+	return true
+}
+
+// called by owner of reconnect state to tell that it has finished reconnecting
+func (client *Client) resetReconnectState() {
+	client.reconnectingActive = false
+}
 
-	// TODO I doubt this error handling is right because it looks
-	// really complicated.
+func (client *Client) reconnect(err error) error {
+
+	// not actioning on error if it's deemed Temporary
+	// we might want to take note of timestamp and eventually recycle this connection
+	// if it persists too long (even though classified as Temporary here)
 	if opErr, ok := err.(*net.OpError); ok {
-		if opErr.Timeout() {
-			client.err(err)
-		}
 		if opErr.Temporary() {
 			return nil
 		}
-
-		return err
 	}
 
-	if err != nil {
-		client.err(err)
+	ownReconnect := client.grabReconnectState()
+
+	if !ownReconnect {
+		//Reconnect collision, this thread will exit and wait on next client.Lock() for other to complete reconnection
+		return nil
 	}
 
-	// If it is unexpected error and the connection wasn't
-	// closed by Gearmand, the client should close the conection
-	// and reconnect to job server.
-	client.Close()
+	defer client.resetReconnectState() // before releasing client lock we will reset reconnection state
+
+	connVersion := client.conn.connVersion
+	close(client.channels.expected)
+	close(client.channels.outbound)
 
-	conn, err := net.Dial(client.net, client.addr)
+	client.Log(Error, fmt.Sprintf("Closing connection to %v due to error %v, will reconnect...", client.addr, err))
+	if close_err := client.Close(); close_err != nil {
+		client.Log(Warning, fmt.Sprintf("Non-fatal error %v, while closing connection to %v", close_err, client.addr))
+	}
 
+	conn, err := client.connOpenHandler()
 	if err != nil {
 		client.err(err)
 		return err
 	}
 
-	swapped := atomic.CompareAndSwapPointer(
-		(*unsafe.Pointer)(unsafe.Pointer(&client.conn)),
-		unsafe.Pointer(nil),
-		unsafe.Pointer(&connection{conn}))
+	newConn := &connection{conn, connVersion + 1}
 
-	if !swapped {
-		conn.Close()
+	if swapped := atomic.CompareAndSwapPointer(
+		(*unsafe.Pointer)(unsafe.Pointer(&client.conn)),
+		unsafe.Pointer(nil), unsafe.Pointer(newConn)); !swapped {
+		return errors.New("Was expecting nil when replacing with new connection")
 	}
 
+	_ = atomic.SwapPointer((*unsafe.Pointer)(unsafe.Pointer(&client.channels)),
```

Comment:
I am not sure this `SwapPointer` is doing much here.

Also, since it always creates a new `chanStruct` when the `connection` is replaced, why not just place the channels in the `connection` object?